### PR TITLE
ENH: weighted metrics

### PR DESCRIPTION
**Pull request by apbard**
I have extracted the distance portion of #6931 and integrated with #6974

Rebased on #7629

### CODE REVIEW

**Comment by rgommers** [comment]
@apbard could you rebase this?

**Comment by apbard** [comment]
@rgommers rebase done

**Comment by rgommers** [comment]
Rebased again to get rid of the test failures (they looked unrelated).

Summary of current changes:

- Weights are added to add distance functions. A lot of detail on definition/usage of weights in https://github.com/scipy/scipy/pull/6931#issuecomment-270817843
- Addresses issue with ``wminkowski`` definition in gh-5718 by deprecating it. No longer needed because ``minkowski`` gains weights.
- Addresses issue with completely dissimilar vectors in ``yule``, ``dice`` and ``sokalsneath``, see gh-1403 and https://github.com/scipy/scipy/pull/6931#issuecomment-270965356. Now returns ``inf`` rather than getting an exception.
- According to the PR description in gh-6931, four functions do not follow all weight invariant operations: ``minkowski``, ``euclidean``, ``sqeuclidean`` and ``canberra``. Exceptions are noted as disabling certain invariant operations in tests.
- Deprecates ``matching`` because it's an alias of ``hamming``.

**Comment by rgommers** [comment]
The deprecations should probably be docstring-only, at least ``wminkowski``. Otherwise it requires scikit-learn to implement awkward logic based on scipy version number.

File: scipy/spatial/distance.py [hunk 1]
```diff
@@ -887,20 +1015,23 @@ def yule(u, v):
...
-    return float(2.0 * ntf * nft) / float(ntt * nff + ntf * nft)
-
-
-def matching(u, v):
+    if w is not None:
+        w = _validate_vector(w)
+    (nff, nft, ntf, ntt) = _nbool_correspond_all(u, v, w=w)
+    return float(2.0 * ntf * nft / np.array(ntt * nff + ntf * nft))
+
+@np.deprecate(message="spatial.distance.matching is deprecated in scipy 0.18.0; "
```
Comment (rgommers) [inline scipy/spatial/distance.py]: If we decide to keep the hard deprecation, then this needs to change to say 1.0

**Comment by rgommers** [comment]
Tests look okay now except for minor issues in a couple of docstrings.

**Comment by apbard** [comment]
@rgommers tests report also 
```
ERROR: missing objects::
    matching
    wminkowski
```
and I don't know how to solve it.
Let me know what you would like me to do for this and the other comments.

**Comment by rgommers** [comment]
@apbard I pushed a fix for all those errors, and changed the deprecation for ``wminkowski`` to be a documentation-only deprecation.

**Comment by rgommers** [comment]
That ``ERROR: missing objects::`` thing you can reproduce locally next time by running ``tools/refguide_check.py``. It checks that all non-deprecated public functions and objects are listed in the module docstring.

**Comment by rgommers** [comment]
> Let me know what you would like me to do for this and the other comments.

Nothing to do right now I think. I refreshed my memory and looked at the big picture. Now need to go through in detail one last time, but I think it's about ready to go in.

File: scipy/spatial/distance.py [hunk 2]
```diff
@@ -1157,7 +1348,7 @@ def sokalsneath(u, v):
 _TEST_METRICS = {'test_' + name: globals()[name] for name in _METRICS.keys()}
 
 
-def pdist(X, metric='euclidean', p=None, w=None, V=None, VI=None):
+def pdist(X, metric='euclidean', *args, **kwargs):
```
Comment (rgommers) [inline scipy/spatial/distance.py]: This is annoying, I thought this PR and gh-7629 were orthogonal. Instead, this is a superset of gh-7629. Let's deal with that one first then.

**Comment by rgommers** [comment]
@apbard could you rebase this on top of master now that the other PR is merged?

Also, 2 test failures in ``test_pdist_calling_conventions`` showed up, can you have a look at that?

**Comment by apbard** [comment]
@rgommers the tests failures are probably related to the [wminkowski deprecation](https://github.com/scipy/scipy/pull/7630/commits/7db7289da127fcf97c5ee61cb6497bad69b86b7d#diff-8ccd60f57f18763fe1d5c56097b49decL480). I'll look into it.

**Comment by rgommers** [comment]
Thanks @apbard. And thanks @metaperture!

**Comment by apbard** [comment]
@rgommers it seems that in one of the rebases I lost the reference to @metaperture authorship. I had to change some stuff but anyway a good portion of the first commit (e5b8e81) it is taken from his PR. Could you please fix the authorship or add a mention in the thanks file?

**Comment by rgommers** [comment]
> Could you please fix the authorship or add a mention in the thanks file?

Can't change the actual commits anymore, but I can add mentions in both THANKS.txt and the 1.0.0 release notes. Would need a full name though. @metaperture could you tell me the name you want to be acknowledged under (either here or email me at ralf.gommers@gmail.com).

**Comment by metaperture** [comment]
Hi @rgommers, I just noticed all this had gotten merged in and tracked this down. My full name is Michael Tartre, and this contribution was made sponsored by / during the course of my work at Two Sigma Investments, LLC. I don't work there any more, but please let me know when you have this attribution up so I can send the reference to my previous team.
Thanks!

**Comment by rgommers** [comment]
missed this first time around, but attribution now updated in gh-8628

**Pull request merged**